be/src/exec/scan_node.cpp

```diff
@@ -63,7 +81,15 @@ StatusOr<pipeline::MorselQueueFactoryPtr> ScanNode::convert_scan_range_to_morsel
                                                     global_scan_ranges, node_id, pipeline_dop,
                                                     enable_tablet_internal_parallel, global_scan_ranges.size()));
         int scan_dop = std::min<int>(std::max<int>(1, morsel_queue->max_degree_of_parallelism()), pipeline_dop);
-        return std::make_unique<pipeline::SharedMorselQueueFactory>(std::move(morsel_queue), scan_dop);
+        int io_parallelism = scan_dop * io_tasks_per_scan_operator();
+
+        // If not so much morsels, try to assign morsel uniformly among operators to avoid data skew
+        if (scan_dop > 1 && morsel_queue->num_original_morsels() <= io_parallelism) {
+            auto morsel_queue_map = uniform_distribute_morsels(std::move(morsel_queue), scan_dop);
+            return std::make_unique<pipeline::IndividualMorselQueueFactory>(std::move(morsel_queue_map));
```

Comment:
Only `FixedMorselQueue` need `uniform_distribute_morsels`, but `morsel_queue` here may be `PhysicalSplitMorselQueue` or `LogicalSplitMorselQueue`.

Reply:
fixed